### ISSUE TITLE
Poor media library performance with large attachment count

### ISSUE THREAD

**Issue by superbia** [deliciousbrains/wp-amazon-s3-and-cloudfront#353]
I have approximately 30,000 media files and load times in the media library are around 16 seconds per page of attachments making it difficult to use. Files are being served from CloudFront and l have the Remove Files From Server option turned on.

The slow down appears to be caused by the [get_attached_file filter](https://github.com/deliciousbrains/wp-amazon-s3-and-cloudfront/blob/master/classes/amazon-s3-and-cloudfront.php#L144) which checks if the file exists locally and appears to run multiple times for each file. Commenting out the filter results in normal load times and no apparent loss of functionality. Is the file check needed if files are being served from S3/CloudFront and the Remove Files From Server option is turned on?

**Comment by ianmjones** [comment]
Hi @superbia,

Please can you post your Diagnostic Info from the Support tab.

**Comment by superbia** [comment]
Thanks Ian, I've emailed you the details.

**Comment by ianmjones** [comment]
`get_attached_file` is specifically for getting a handle to a local version of the file, generally because a plugin wants do some work with it rather than just display it. Therefore if the file does exist locally, job's done, we hand that back. Otherwise, if "Remove Files From Server" is turned on and we have the file on S3 we can either hand back a stream-wrapper handle or pull back the file and hand that back (if the appropriate filter is used).

This is very different to the usual `wp_get_attachment_url` that is used for displaying attachments, and is generally used by the Media Library list.

You really don't want to disable the `get_attached_file` filter.

I would recommend experimenting by re-enabling the filter, then going through other plugins one by one temporarily deactivating them and checking whether you get a speedup in the Media Library after each deactivation. Your diagnostics had a couple of MU plugins that are potential candidates, your others aren't anything out of the ordinary but worth checking.

If you have anything in your theme's function.php file, that might be worth investigating.

When the Media Library list is displayed, do the thumbnails have local or S3 URLs?

**Comment by superbia** [comment]
Thanks for getting back to me so quickly. I disabled all of the plugins including the must-use plugins and removed the functionality in my theme's functions.php and it seems that there's a conflict with Advanced Custom Fields Pro. I originally disabled the plugins I thought would be an issue and then started looking at the filters in WP Offload S3 Lite... sorry I should have gone through the plugins more thoroughly.

To clarify the problem, requests to admin-ajax.php are taking over 10 seconds in the media library. If I disable Advanced Custom Fields the attachments load at normal speed and the thumbnails have S3 urls. If I keep ACF Pro enabled but disable WP Offload S3 Lite the attachments load at normal speed as well but the thumbnails 404 as they're only hosted on S3.

Cheers,

Dylan

**Comment by ianmjones** [comment]
That's weird, I can't think of any reason that ACF Pro would even load on the Media Library page.

Thanks for digging in and finding the issue.

**Comment by superbia** [comment]
I tracked this down to the ACF file field getting the file size of attachments on the `wp_prepare_attachment_for_js` filter. The`get_attached_file` function is returning s3 protocol urls and then the filesize function is actually returning the correct file size of the remote assets on s3 which is what is slowing everything down. Should I talk to Elliot over at ACF about this or is it a joint issue?

**Comment by ianmjones** [comment]
I think you should ask the ACF developers about this, as it seems a little strange for that plugin to be doing that kind of work on the Media page.

We do have a workaround, there's a `as3cf_get_attached_file_copy_back_to_local ` filter you can return `true` from when you need to have a local file available because a plugin doesn't work well with remote media.

https://github.com/deliciousbrains/wp-amazon-s3-and-cloudfront-tweaks/blob/master/amazon-s3-and-cloudfront-tweaks.php#L236

**Comment by tunelinks** [comment]
@superbia I'm running into the same issue.  I'm getting major performance issues after enabling this plugin and I use ACF as well.

What was your solution?

**Comment by arwinjp** [comment]
Media Search is taking lot of time to search the media that was offloaded to DO Spaces. I have about 170GB of media files. Is there any way to speed up the search process?
Note that I have opted to remove files from local folder, all the files are available only in cloud. Any help is really appreciated.
